tests/test_feedexport.py

```diff
@@ -1320,6 +1320,76 @@ def test_extend_kwargs(self):
             data = yield self.exported_data(items, settings)
             self.assertEqual(row['expected'], data[feed_options['format']])
 
+    @defer.inlineCallbacks
+    def test_postprocessed_exports(self):
+        items = [
+            {'foo': 'bar1', 'baz': ''},
+            {'egg': 'bar2', 'baz': 'quux'},
+        ]
+
+        import gzip
+        import lzma
+        import bz2
+        format_to_decompressor = {
+            'csv': gzip,
+            'json': lzma,
+            'jsonlines': bz2,
+            'xml': gzip,
+        }
+
+        format_to_expected = {
+            'csv': b'foo,baz\r\nbar1,\r\n,quux\r\n',
+            'json': b'[\n{"foo": "bar1", "baz": ""},\n{"egg": "bar2", "baz": "quux"}\n]',
+            'jsonlines': b'{"foo": "bar1", "baz": ""}\n{"egg": "bar2", "baz": "quux"}\n',
+            'xml': b'XXXX<?xml version="1.0" encoding="utf-8"?>\nXXXX<itemsXXXX>XXXX\nX'
+                   b'XXX<itemXXXX>XXXX<fooXXXX>XXXXbar1XXXX</foo>XXXX<bazXXXX>XXXX</baz'
+                   b'>XXXX</item>XXXX\nXXXX<itemXXXX>XXXX<eggXXXX>XXXXbar2XXXX</egg>XXX'
+                   b'X<bazXXXX>XXXXquuxXXXX</baz>XXXX</item>XXXX\nXXXX</items>',
+        }
+
+        class MyPlugin1:
+            def __init__(self, file, feed_options):
+                self.file = file
+                self.feed_options = feed_options
+
+            def write(self, data):
+                return self.file.write(b"XXXX" + data)
+
+            def close(self):
+                self.file.close()
+
+        settings = {
+            'FEEDS': {
+                self._random_temp_filename(): {
+                    'format': 'csv',
+                    'postprocessing': ['scrapy.extensions.postprocessing.GzipPlugin'],
+                    'gzip_compresslevel': random.randint(1, 9),
+                },
+                self._random_temp_filename(): {
+                    'format': 'json',
+                    'postprocessing': ['scrapy.extensions.postprocessing.LZMAPlugin'],
+                    'lzma_preset': random.randint(0, 9),
+                    'lzma_check': random.choice([lzma.CHECK_CRC32, lzma.CHECK_CRC64,
+                                                 lzma.CHECK_NONE, lzma.CHECK_SHA256]),
+                },
+                self._random_temp_filename(): {
+                    'format': 'jsonlines',
+                    'postprocessing': ['scrapy.extensions.postprocessing.Bz2Plugin'],
+                    'bz2_compresslevel': random.randint(1, 9),
+                },
+                self._random_temp_filename(): {
+                    'format': 'xml',
+                    'postprocessing': [MyPlugin1, 'scrapy.extensions.postprocessing.GzipPlugin'],
+                },
+            }
+        }
+
+        data = yield self.exported_data(items, settings)
+
+        for fmt, decompressor in format_to_decompressor.items():
+            result = decompressor.decompress(data[fmt])
+            self.assertEqual(format_to_expected[fmt], result)
```

Comment:
The test looks great overall. It should serve to make sure that all built-in compression plugins work.

They do not hurt, but I don’t think the (random) extra parameters add much to the test. For example, if you set `gzip_compresslevel` to `4` in this test, but the plugin were not processing the parameter at all, the test would still pass. It may be better to remove them.

To test plugin parameters, you need tests that you break if parameters were misinterpreted. You want to make sure that the parameters have an impact on the output, that the output is not the same with different parameters. You don’t need to test the underlying library (e.g. you don’t need to test that Python’s GZipFile makes files smaller with 9 than with 0), but you do need to test somehow that the plugin is passing the parameter (e.g. from feed options to GZipFile).

You also need to mind that some compressors may not be deterministic (e.g. [GZipFile seems to not be deterministic unless you set `mtime` to a fixed value](https://discuss.python.org/t/gzip-py-allow-deterministic-compression-without-time-stamp/8221)), so it may not always be enough to test that the output is different with a different value in a parameter. In these cases, you’ll need to either make the output deterministic or check for something more. For example, with GZipFile, you can either expose the `mtime` option and use it in tests in combination with the options that you are actually testing, or you can test an option like compression level by verifying that the higher the compression the smaller the file.